File: .github/scripts/push_package_tree.py
#!/usr/bin/env python3
"""
    Assembles the package file tree on the shim fork via the GitHub Git Data API.

Reads files from packages/{package_name}/, respects the `exclude` globs in
typst.toml, creates blobs on the fork, creates a tree + commit, and advances
the branch ref.  No local git operations — pure REST API.
"""

import argparse
import base64
import json
import subprocess
import sys
import tomllib
from fnmatch import fnmatch
from pathlib import Path


def api(method: str, path: str, payload: dict | None = None) -> dict:
    cmd = ["gh", "api", path, "-X", method]
    if payload is not None:
        result = subprocess.run(
            cmd + ["--input", "-"],
            input=json.dumps(payload),
            capture_output=True,
            text=True,
        )
    else:
        result = subprocess.run(cmd, capture_output=True, text=True)

    if result.returncode != 0:
        print(f"::error::API error ({method} {path}):\n{result.stderr}", file=sys.stderr)
        sys.exit(1)

    return json.loads(result.stdout)


def get_exclude_patterns(package_dir: Path) -> list[str]:
    with open(package_dir / "typst.toml", "rb") as f:
        config = tomllib.load(f)
    return config.get("package", {}).get("exclude", [])


def is_excluded(rel: Path, patterns: list[str]) -> bool:
    rel_str = str(rel)
    for pattern in patterns:
        if fnmatch(rel_str, pattern):
            return True
        if fnmatch(rel.name, pattern):
            return True
        # Also match any intermediate directory segment
        for part in rel.parts:
            if fnmatch(part, pattern):
                return True
    return False


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--package-name", required=True)
    parser.add_argument("--version", required=True)
    parser.add_argument("--branch", required=True)
    parser.add_argument("--fork", required=True)
    args = parser.parse_args()

    package_dir = Path("packages") / args.package_name
    if not package_dir.is_dir():
        print(f"::error::Package directory not found: {package_dir}", file=sys.stderr)
        sys.exit(1)

    exclude_patterns = get_exclude_patterns(package_dir)
    dest_prefix = f"packages/preview/{args.package_name}/{args.version}"

    # Resolve the branch's current commit and tree SHA
    ref_data = api("GET", f"repos/{args.fork}/git/ref/heads/{args.branch}")
    commit_sha: str = ref_data["object"]["sha"]
    commit_data = api("GET", f"repos/{args.fork}/git/commits/{commit_sha}")
    base_tree_sha: str = commit_data["tree"]["sha"]

    print(f"Base commit : {commit_sha}")
    print(f"Base tree   : {base_tree_sha}")

    # Collect files, applying exclude patterns
    files: list[tuple[Path, Path]] = []
    for path in sorted(package_dir.rglob("*")):
        if not path.is_file():
            continue
        rel = path.relative_to(package_dir)
        if is_excluded(rel, exclude_patterns):
            print(f"  (excluded) {rel}")
            continue
        files.append((path, rel))

    print(f"\nAssembling {len(files)} file(s) → {dest_prefix}/")

    # Create a blob for every file
    tree_entries: list[dict] = []
    for abs_path, rel in files:
        content_b64 = base64.b64encode(abs_path.read_bytes()).decode()
        blob = api(
            "POST",
            f"repos/{args.fork}/git/blobs",
            {"content": content_b64, "encoding": "base64"},
        )
        dest_path = f"{dest_prefix}/{rel}"
        tree_entries.append(
            {"path": dest_path, "mode": "100644", "type": "blob", "sha": blob["sha"]}
        )
        print(f"  + {dest_path}")

    # Create the tree on top of the fork's current HEAD tree
    tree = api(
        "POST",
        f"repos/{args.fork}/git/trees",
        {"base_tree": base_tree_sha, "tree": tree_entries},
    )
    print(f"\nCreated tree   : {tree['sha']}")

    # Create the commit
    commit = api(
        "POST",
        f"repos/{args.fork}/git/commits",
        {
            "message": f"packages/preview/{args.package_name}/{args.version}",
            "tree": tree["sha"],
            "parents": [commit_sha],
        },
    )
    print(f"Created commit : {commit['sha']}")

    # Advance the branch ref (fast-forward only)
    api(
        "PATCH",
        f"repos/{args.fork}/git/refs/heads/{args.branch}",
        {"sha": commit["sha"], "force": False},
    )
    print(f"Branch {args.branch} → {commit['sha']}")


if __name__ == "__main__":
    main()

File: .github/scripts/push-package-tree.js
#!/usr/bin/env node
'use strict';

/**
 * Assembles the package file tree on the shim fork via the GitHub Git Data API.
 *
 * Reads files from packages/{packageName}/, respects the `exclude` globs in
 * typst.toml, creates blobs on the fork, creates a tree + commit, and advances
 * the branch ref.  No local git operations — pure REST API via `gh` CLI.
 *
 * Usage:
 *   node push-package-tree.js --package-name <name> --version <ver> \
 *                             --branch <branch> --fork <owner/repo>
 */

const { spawnSync } = require('node:child_process');
const { readFileSync, readdirSync } = require('node:fs');
const path = require('node:path');

// ── Helpers ───────────────────────────────────────────────────────────────────

function ghApi(method, apiPath, payload) {
  const args = ['api', apiPath, '-X', method];
  const opts = { encoding: 'utf8', maxBuffer: 64 * 1024 * 1024 };
  if (payload != null) {
    args.push('--input', '-');
    opts.input = JSON.stringify(payload);
  }
  const result = spawnSync('gh', args, opts);
  if (result.status !== 0) {
    console.error(`::error::API error (${method} ${apiPath}):\n${result.stderr}`);
    process.exit(1);
  }
  return JSON.parse(result.stdout);
}

function parseArgs() {
  const args = process.argv.slice(2);
  const out = {};
  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace(/^--/, '').replace(/-([a-z])/g, (_, c) => c.toUpperCase());
    out[key] = args[i + 1];
  }
  return out;
}

function getExcludePatterns(packageDir) {
  const toml = readFileSync(path.join(packageDir, 'typst.toml'), 'utf8');
  // Handles both single-line and multiline arrays
  const match = toml.match(/exclude\s*=\s*\[([^\]]*)\]/s);
  if (!match) return [];
  return match[1]
    .split(',')
    .map(s => s.trim().replace(/^["']|["']$/g, ''))
    .filter(Boolean);
}

function globToRegex(pattern) {
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
    .replace(/\*\*/g, '\x00')
    .replace(/\*/g, '[^/]*')
    .replace(/\?/g, '[^/]')
    .replace(/\x00/g, '.*');
  return new RegExp(`^${escaped}$`);
}

function isExcluded(rel, patterns) {
  const relPosix = rel.split(path.sep).join('/');
  const parts = relPosix.split('/');
  for (const pattern of patterns) {
    const re = globToRegex(pattern);
    if (re.test(relPosix)) return true;
    if (re.test(parts[parts.length - 1])) return true;
    for (const part of parts) {
      if (re.test(part)) return true;
    }
  }
  return false;
}

function* walkDir(dir) {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) yield* walkDir(full);
    else if (entry.isFile()) yield full;
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────

function main() {
  const { packageName, version, branch, fork } = parseArgs();
  const packageDir = path.join('packages', packageName);
  const destPrefix = `packages/preview/${packageName}/${version}`;

  // Resolve the branch's current commit and tree SHA
  const refData = ghApi('GET', `repos/${fork}/git/ref/heads/${branch}`);
  const commitSha = refData.object.sha;
  const commitData = ghApi('GET', `repos/${fork}/git/commits/${commitSha}`);
  const baseTreeSha = commitData.tree.sha;

  console.log(`Base commit : ${commitSha}`);
  console.log(`Base tree   : ${baseTreeSha}`);

  const excludePatterns = getExcludePatterns(packageDir);

  // Collect files, applying exclude patterns
  const files = [];
  for (const absPath of walkDir(packageDir)) {
    const rel = path.relative(packageDir, absPath);
    if (isExcluded(rel, excludePatterns)) {
      console.log(`  (excluded) ${rel}`);
      continue;
    }
    files.push({ absPath, rel });
  }
  files.sort((a, b) => a.rel.localeCompare(b.rel));

  console.log(`\nAssembling ${files.length} file(s) → ${destPrefix}/`);

  // Create a blob for every file
  const treeEntries = [];
  for (const { absPath, rel } of files) {
    const contentB64 = readFileSync(absPath).toString('base64');
    const blob = ghApi('POST', `repos/${fork}/git/blobs`, {
      content: contentB64,
      encoding: 'base64',
    });
    const destPath = `${destPrefix}/${rel.split(path.sep).join('/')}`;
    treeEntries.push({ path: destPath, mode: '100644', type: 'blob', sha: blob.sha });
    console.log(`  + ${destPath}`);
  }

  // Create tree on top of fork's current HEAD tree
  const tree = ghApi('POST', `repos/${fork}/git/trees`, {
    base_tree: baseTreeSha,
    tree: treeEntries,
  });
  console.log(`\nCreated tree   : ${tree.sha}`);

  // Create the commit
  const commit = ghApi('POST', `repos/${fork}/git/commits`, {
    message: `packages/preview/${packageName}/${version}`,
    tree: tree.sha,
    parents: [commitSha],
  });
  console.log(`Created commit : ${commit.sha}`);

  // Advance the branch ref (fast-forward only)
  ghApi('PATCH', `repos/${fork}/git/refs/heads/${branch}`, {
    sha: commit.sha,
    force: false,
  });
  console.log(`Branch ${branch} → ${commit.sha}`);
}

main();
